File: backend/integrations.py
"""External integrations: WhatsApp, email, HMAC-signed CRM webhook, IndexNow, Google Search Console."""
from typing import Optional, List
import hashlib
import hmac
import json
import time
import uuid
import requests
import urllib.parse

from deps import (
    TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_WHATSAPP_FROM,
    RESEND_API_KEY, RESEND_FROM_EMAIL, ADMIN_NOTIFICATION_EMAILS,
    WEBHOOK_URL, WEBHOOK_SECRET,
    INDEXNOW_KEY, INDEXNOW_ENDPOINT,
    GSC_SERVICE_ACCOUNT_JSON,
    logger,
)


# -------------------- WhatsApp --------------------
def send_whatsapp(lead: dict):
    if not (TWILIO_ACCOUNT_SID and TWILIO_AUTH_TOKEN and TWILIO_WHATSAPP_FROM):
        logger.info("[STUB] WhatsApp not configured — would notify %s", lead.get("whatsapp_number"))
        return
    try:
        from twilio.rest import Client
        c = Client(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN)
        body = (
            f"Hi {lead.get('first_name')}, this is AussieBack. We've received your estimate "
            f"of ${lead.get('estimated_refund'):,.0f}. Our team will contact you shortly to "
            f"fast-track your DASP refund. Reply STOP to opt-out."
        )
        to = lead.get("whatsapp_number", "")
        if not to.startswith("whatsapp:"):
            to = f"whatsapp:{to}"
        msg = c.messages.create(body=body, from_=TWILIO_WHATSAPP_FROM, to=to)
        logger.info("WhatsApp sent SID=%s", msg.sid)
    except Exception as e:
        logger.exception("WhatsApp send failed: %s", e)


# -------------------- Email --------------------
def send_emails(lead: dict):
    if not (RESEND_API_KEY and RESEND_FROM_EMAIL):
        logger.info("[STUB] Email not configured — would email %s", lead.get("email"))
        return
    try:
        import resend
        resend.api_key = RESEND_API_KEY
        resend.Emails.send({
            "from": RESEND_FROM_EMAIL,
            "to": [lead["email"]],
            "subject": f"Your AussieBack Super refund estimate: ${lead['estimated_refund']:,.0f}",
            "html": f"""
                <h2>Hi {lead['first_name']},</h2>
                <p>Thanks for using AussieBack. Based on what you've told us, your estimated
                Australian Super refund is <strong>${lead['estimated_refund']:,.0f}</strong>.</p>
                <p>Our team will reach out via WhatsApp shortly to help you recover this.</p>
                <p>– The AussieBack Team</p>
            """,
        })
        admin_recipients = [e.strip() for e in ADMIN_NOTIFICATION_EMAILS.split(",") if e.strip()]
        if admin_recipients:
            resend.Emails.send({
                "from": RESEND_FROM_EMAIL,
                "to": admin_recipients,
                "subject": f"New AussieBack lead: {lead['first_name']} (${lead['estimated_refund']:,.0f})",
                "html": f"""
                    <h3>New lead captured</h3>
                    <ul>
                      <li>Name: {lead['first_name']}</li>
                      <li>Email: {lead['email']}</li>
                      <li>WhatsApp: {lead['whatsapp_number']}</li>
                      <li>Visa: {lead['visa_type']}</li>
                      <li>Estimated refund: ${lead['estimated_refund']:,.0f}</li>
                      <li>Super fund: {lead.get('super_fund_name') or 'N/A'}</li>
                      <li>Departure: {lead.get('date_left_australia') or 'N/A'}</li>
                    </ul>
                """,
            })
    except Exception as e:
        logger.exception("Resend email failed: %s", e)


# -------------------- HMAC-signed CRM webhook (Zapier / Zoho / Hubspot) --------------------
def _sign_payload(payload: bytes) -> str:
    """SHA-256 HMAC in hex, prefixed with 'sha256=' to match GitHub/Stripe convention."""
    if not WEBHOOK_SECRET:
        return ""
    digest = hmac.new(WEBHOOK_SECRET.encode("utf-8"), payload, hashlib.sha256).hexdigest()
    return f"sha256={digest}"


def send_webhook(event: str, data: dict, *, previous: Optional[dict] = None) -> None:
    """Forward a generic event to the configured webhook (fire-and-forget).

    Payload shape (JSON):
        {
          "event": "lead.created" | "lead.status_changed" | "comment.created" | "share_event.created",
          "id": "<uuid>",
          "occurred_at": "2026-...Z",
          "data": {...},        # the object itself
          "previous": {...}     # for status changes only — the old subset
        }

    Signature: sha256=<hex> HMAC over the raw body, sent as `X-AussieBack-Signature`.
    """
    if not WEBHOOK_URL:
        logger.info("[STUB] Webhook not configured — would forward event=%s id=%s", event, data.get("id"))
        return
    envelope = {
        "event": event,
        "id": str(uuid.uuid4()),
        "occurred_at": _now_iso(),
        "data": data,
    }
    if previous is not None:
        envelope["previous"] = previous
    body = json.dumps(envelope, default=str).encode("utf-8")
    headers = {
        "Content-Type": "application/json",
        "User-Agent": "AussieBack-Webhook/1.0",
        "X-AussieBack-Event": event,
    }
    sig = _sign_payload(body)
    if sig:
        headers["X-AussieBack-Signature"] = sig
    try:
        resp = requests.post(WEBHOOK_URL, data=body, headers=headers, timeout=(3, 10))
        logger.info("Webhook POST event=%s status=%s", event, resp.status_code)
    except Exception as e:
        logger.exception("Webhook POST failed (event=%s): %s", event, e)


def _now_iso() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()


def dispatch_lead_integrations(lead: dict):
    """Fired once when a lead is created."""
    send_whatsapp(lead)
    send_emails(lead)
    send_webhook("lead.created", lead)


# -------------------- IndexNow (Bing/Yandex/DuckDuckGo instant re-crawl) --------------------
def ping_indexnow(urls: List[str], host: str) -> Optional[dict]:
    """POST a list of URLs to the IndexNow endpoint.

    Requires `INDEXNOW_KEY` env var (a random 8–128 char hex string) and a
    key-verification file served at `https://<host>/<INDEXNOW_KEY>.txt`.
    The backend exposes `GET /{key}.txt` via `routes.seo` so this works out of the box.

    Returns None (no-op) when INDEXNOW_KEY is unset.
    """
    if not INDEXNOW_KEY or not urls:
        if not INDEXNOW_KEY:
            logger.info("[STUB] IndexNow not configured — would ping %d urls", len(urls))
        return None
    parsed_host = urllib.parse.urlparse(host).netloc or host
    payload = {
        "host": parsed_host,
        "key": INDEXNOW_KEY,
        "keyLocation": f"{host.rstrip('/')}/{INDEXNOW_KEY}.txt",
        "urlList": urls,
    }
    try:
        r = requests.post(
            INDEXNOW_ENDPOINT,
            json=payload,
            headers={"Content-Type": "application/json; charset=utf-8"},
            timeout=(3, 10),
        )
        logger.info("IndexNow ping status=%s urls=%d host=%s", r.status_code, len(urls), parsed_host)
        return {"status": r.status_code, "urls": len(urls)}
    except Exception as e:
        logger.exception("IndexNow ping failed: %s", e)
        return {"error": str(e), "urls": len(urls)}


# -------------------- Google Search Console API (optional, OAuth service account) --------------------
def ping_gsc_sitemap(sitemap_url: str, site_property: str) -> Optional[dict]:
    """Force a sitemap re-submission via the Search Console API.

    Requires `GSC_SERVICE_ACCOUNT_JSON` env var containing the entire service-
    account JSON string. The service account email must be added as an owner
    for `site_property` in the Search Console UI. `site_property` should look
    like "https://aussieback.com/" (with trailing slash) or "sc-domain:aussieback.com".

    Returns None (no-op) when unconfigured.
    """
    if not GSC_SERVICE_ACCOUNT_JSON:
        logger.info("[STUB] GSC_SERVICE_ACCOUNT_JSON not configured — skipping GSC ping")
        return None
    try:
        # Lazy imports so the base install stays lean when GSC is unused.
        from google.oauth2 import service_account
        from googleapiclient.discovery import build
    except ImportError:
        logger.warning("google-api-python-client / google-auth not installed — cannot ping GSC")
        return {"error": "missing_google_client"}
    try:
        info = json.loads(GSC_SERVICE_ACCOUNT_JSON)
        creds = service_account.Credentials.from_service_account_info(
            info, scopes=["https://www.googleapis.com/auth/webmasters"]
        )
        service = build("searchconsole", "v1", credentials=creds, cache_discovery=False)
        # `submit` on `sitemaps` triggers Google to re-fetch immediately.
        service.sitemaps().submit(siteUrl=site_property, feedpath=sitemap_url).execute()
        logger.info("GSC sitemap re-submitted: %s (property=%s)", sitemap_url, site_property)
        return {"ok": True, "sitemap": sitemap_url}
    except Exception as e:
        logger.exception("GSC ping failed: %s", e)
        return {"error": str(e)}


def ping_search_engines_for_slug(site_url: str, slugs: List[str]) -> dict:
    """Convenience: build blog URLs from slugs and ping IndexNow + GSC in one call.

    site_url is expected to be the canonical origin (e.g. "https://aussieback.com").
    """
    urls = [f"{site_url.rstrip('/')}/blog/{s}" for s in slugs if s]
    # Always include /blog itself so listing pages get re-crawled too.
    urls.append(f"{site_url.rstrip('/')}/blog")
    urls.append(f"{site_url.rstrip('/')}/sitemap.xml")
    indexnow_result = ping_indexnow(urls, site_url)
    gsc_result = ping_gsc_sitemap(
        f"{site_url.rstrip('/')}/sitemap.xml",
        f"{site_url.rstrip('/')}/",
    )
    return {"indexnow": indexnow_result, "gsc": gsc_result, "urls": urls}

File: src/lib/server/integrations.js
// External integrations: WhatsApp, email, HMAC-signed CRM webhook, IndexNow,
// Google Search Console. Port of backend/integrations.py — WhatsApp now uses
// the `twilio` npm package (same Twilio API), email goes through mailer.js
// (Nodemailer/SMTP replacing Resend), and the webhook path is unchanged
// (enqueue into the durable outbox, HMAC-signed on delivery by outbox.js).
import twilio from "twilio";
import { sendMail } from "./mailer";
import { enqueue } from "./outbox";
import { renderEmail, button, escapeHtml, BRAND, SITE_URL } from "./emailTemplates";

const VISA_LABEL = { working_holiday: "Working Holiday Maker", other_temp: "Student / Other Temporary Visa" };
const fmtAUD = (n) => `$${Math.round(Number(n) || 0).toLocaleString()}`;

// -------------------- WhatsApp --------------------
export async function sendWhatsapp(lead) {
  const { TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN, TWILIO_WHATSAPP_FROM } = process.env;
  if (!(TWILIO_ACCOUNT_SID && TWILIO_AUTH_TOKEN && TWILIO_WHATSAPP_FROM)) {
    console.log(`[STUB] WhatsApp not configured — would notify ${lead.whatsapp_number}`);
    return;
  }
  try {
    const client = twilio(TWILIO_ACCOUNT_SID, TWILIO_AUTH_TOKEN);
    const body =
      `Hi ${lead.first_name}, this is refundmysuper. We've received your estimate ` +
      `of $${Math.round(lead.estimated_refund).toLocaleString()}. Our team will contact you shortly to ` +
      `fast-track your DASP refund. Reply STOP to opt-out.`;
    let to = lead.whatsapp_number || "";
    if (!to.startsWith("whatsapp:")) to = `whatsapp:${to}`;
    const msg = await client.messages.create({ body, from: TWILIO_WHATSAPP_FROM, to });
    console.log(`WhatsApp sent SID=${msg.sid}`);
  } catch (e) {
    console.error("WhatsApp send failed:", e);
  }
}

// -------------------- Email --------------------
export async function sendLeadEmails(lead) {
  if (!process.env.SMTP_HOST) {
    console.log(`[STUB] Email not configured — would email ${lead.email}`);
    return;
  }
  const adminRecipients = (process.env.ADMIN_NOTIFICATION_EMAILS || "")
    .split(",")
    .map((e) => e.trim())
    .filter(Boolean);

  const sends = [
    sendMail({
      to: lead.email,
      subject: `Your refundmysuper Australian Super refund estimate: ${fmtAUD(lead.estimated_refund)}`,
      html: renderEmail({
        preheader: `Your estimated refund is ${fmtAUD(lead.estimated_refund)} — here's what happens next.`,
        bodyHtml: `
          <h1 style="margin:0 0 14px;font-family:Arial,Helvetica,sans-serif;font-size:24px;font-weight:700;color:${BRAND.navy};">
            Hi ${escapeHtml(lead.first_name)}, your super is on its way home.
          </h1>
          <p style="margin:0 0 24px;font-size:15px;line-height:1.6;color:${BRAND.muted};">
            Thanks for using refundmysuper. Based on what you told us, here's your free DASP refund estimate:
          </p>
          <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="background:${BRAND.navy};border-radius:14px;margin-bottom:24px;">
            <tr>
              <td align="center" style="padding:28px 20px;">
                <div style="font-family:Arial,Helvetica,sans-serif;font-size:12px;letter-spacing:0.12em;text-transform:uppercase;color:rgba(255,255,255,0.75);margin-bottom:8px;">
                  Estimated refund
                </div>
                <div style="font-family:Arial,Helvetica,sans-serif;font-size:42px;font-weight:700;color:#FFFFFF;line-height:1.1;">
                  ${fmtAUD(lead.estimated_refund)}
                </div>
                <div style="display:inline-block;margin-top:14px;padding:6px 16px;background:rgba(255,255,255,0.15);border-radius:999px;font-family:Arial,Helvetica,sans-serif;font-size:12px;font-weight:600;color:${BRAND.gold};">
                  ${escapeHtml(VISA_LABEL[lead.visa_type] || "Temporary visa")}
                </div>
              </td>
            </tr>
          </table>
          <h2 style="margin:0 0 12px;font-family:Arial,Helvetica,sans-serif;font-size:16px;color:${BRAND.navy};">What happens next</h2>
          <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="margin-bottom:28px;">
            ${[
              ["1", "We review your details", "Our TPB-partnered tax agents check your eligibility — no paperwork needed from you yet."],
              ["2", "We message you on WhatsApp", `We'll reach out at ${escapeHtml(lead.whatsapp_number || "your WhatsApp number")} within 1 business day.`],
              ["3", "You get paid", "Once lodged, DASP payments usually land within 28 days — straight to your bank."],
            ]
              .map(
                ([n, title, body]) => `
              <tr>
                <td width="36" valign="top" style="padding:0 0 18px;">
                  <table role="presentation" cellpadding="0" cellspacing="0"><tr><td style="width:26px;height:26px;border-radius:50%;background:#EBF3FA;color:${BRAND.navy};font-family:Arial,Helvetica,sans-serif;font-size:13px;font-weight:700;text-align:center;line-height:26px;">${n}</td></tr></table>
                </td>
                <td valign="top" style="padding:0 0 18px 12px;">
                  <div style="font-family:Arial,Helvetica,sans-serif;font-size:14px;font-weight:700;color:${BRAND.ink};margin-bottom:2px;">${escapeHtml(title)}</div>
                  <div style="font-family:Arial,Helvetica,sans-serif;font-size:13px;line-height:1.5;color:${BRAND.muted};">${body}</div>
                </td>
              </tr>`
              )
              .join("")}
          </table>
          <div style="margin-bottom:8px;">
            ${button({ href: `${SITE_URL}/#estimator`, label: "View my estimate again" })}
          </div>
          <p style="margin:24px 0 0;font-size:14px;line-height:1.6;color:${BRAND.muted};">
            Questions in the meantime? Just reply to this email or WhatsApp us — we usually respond within an hour.
          </p>
          <p style="margin:20px 0 0;font-size:14px;color:${BRAND.ink};">– The refundmysuper Team</p>
        `,
      }),
    }),
  ];
  if (adminRecipients.length) {
    const rows = [
      ["Name", escapeHtml(lead.first_name)],
      ["Email", `<a href="mailto:${escapeHtml(lead.email)}" style="color:${BRAND.navy};">${escapeHtml(lead.email)}</a>`],
      ["WhatsApp", escapeHtml(lead.whatsapp_number)],
      ["Visa", escapeHtml(VISA_LABEL[lead.visa_type] || lead.visa_type)],
      ["Estimated refund", `<strong>${fmtAUD(lead.estimated_refund)}</strong>`],
      ["Super fund", escapeHtml(lead.super_fund_name || "N/A")],
      ["Departure date", escapeHtml(lead.date_left_australia || "N/A")],
    ];
    sends.push(
      sendMail({
        to: adminRecipients,
        subject: `New refundmysuper lead: ${lead.first_name} (${fmtAUD(lead.estimated_refund)})`,
        html: renderEmail({
          preheader: `New lead: ${lead.first_name} — ${fmtAUD(lead.estimated_refund)} estimated refund.`,
          bodyHtml: `
            <h1 style="margin:0 0 4px;font-family:Arial,Helvetica,sans-serif;font-size:22px;font-weight:700;color:${BRAND.navy};">New lead captured</h1>
            <p style="margin:0 0 22px;font-size:14px;color:${BRAND.muted};">A visitor just completed a refund estimate.</p>
            <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="border:1px solid ${BRAND.border};border-radius:12px;overflow:hidden;margin-bottom:24px;">
              ${rows
                .map(
                  ([label, value], i) => `
                <tr>
                  <td style="padding:12px 16px;background:${i % 2 === 0 ? "#F8FAFC" : "#FFFFFF"};font-family:Arial,Helvetica,sans-serif;font-size:13px;color:${BRAND.muted};width:150px;border-bottom:1px solid ${BRAND.border};">${label}</td>
                  <td style="padding:12px 16px;background:${i % 2 === 0 ? "#F8FAFC" : "#FFFFFF"};font-family:Arial,Helvetica,sans-serif;font-size:13px;color:${BRAND.ink};border-bottom:1px solid ${BRAND.border};">${value}</td>
                </tr>`
                )
                .join("")}
            </table>
            ${button({ href: `${SITE_URL}/admin`, label: "Open admin dashboard" })}
          `,
        }),
      })
    );
  }
  // Independent recipients — send concurrently rather than serializing two
  // SMTP round-trips on the /api/leads response path.
  await Promise.allSettled(sends);
}

// -------------------- HMAC-signed CRM webhook (Zapier / Zoho / Hubspot) --------------------
/**
 * Enqueue an outbound event into the durable outbox (see outbox.js). When
 * WEBHOOK_URL is unset this is a no-op, mirroring the Python stub behaviour.
 */
export function sendWebhook(event, data, { previous } = {}) {
  if (!process.env.WEBHOOK_URL) {
    console.log(`[STUB] Webhook not configured — would forward event=${event} id=${data?.id}`);
    return Promise.resolve(null);
  }
  return enqueue(event, data, { previous });
}

export async function dispatchLeadIntegrations(lead) {
  // Run concurrently, not sequentially — this sits on the /api/leads response
  // path, so a slow provider (e.g. SMTP) shouldn't add its latency on top of
  // the others'. Each call already catches its own errors internally, but
  // allSettled is a second guard against one hanging integration blocking
  // the others from even starting.
  await Promise.allSettled([sendWhatsapp(lead), sendLeadEmails(lead), sendWebhook("lead.created", lead)]);
}

// -------------------- IndexNow (Bing/Yandex/DuckDuckGo instant re-crawl) --------------------
export async function pingIndexnow(urls, host) {
  const INDEXNOW_KEY = process.env.INDEXNOW_KEY || "";
  const INDEXNOW_ENDPOINT = process.env.INDEXNOW_ENDPOINT || "https://api.indexnow.org/indexnow";
  if (!INDEXNOW_KEY || !urls.length) {
    if (!INDEXNOW_KEY) console.log(`[STUB] IndexNow not configured — would ping ${urls.length} urls`);
    return null;
  }
  let parsedHost;
  try {
    parsedHost = new URL(host).host;
  } catch (e) {
    parsedHost = host;
  }
  const payload = {
    host: parsedHost,
    key: INDEXNOW_KEY,
    keyLocation: `${host.replace(/\/$/, "")}/${INDEXNOW_KEY}.txt`,
    urlList: urls,
  };
  try {
    const resp = await fetch(INDEXNOW_ENDPOINT, {
      method: "POST",
      headers: { "Content-Type": "application/json; charset=utf-8" },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(10000),
    });
    console.log(`IndexNow ping status=${resp.status} urls=${urls.length} host=${parsedHost}`);
    return { status: resp.status, urls: urls.length };
  } catch (e) {
    console.error("IndexNow ping failed:", e);
    return { error: String(e), urls: urls.length };
  }
}

// -------------------- Google Search Console (optional, service-account OAuth) --------------------
/**
 * Force a sitemap re-submission via the Search Console API's REST endpoint.
 * The plan specifies `google-auth-library` (not the full `googleapis` client)
 * for this port, so the sitemap-submit call is made directly against the
 * Webmasters v3 REST API rather than through a generated client — functionally
 * identical to the Python version's `service.sitemaps().submit(...)`.
 */
export async function pingGscSitemap(sitemapUrl, siteProperty) {
  const GSC_SERVICE_ACCOUNT_JSON = process.env.GSC_SERVICE_ACCOUNT_JSON || "";
  if (!GSC_SERVICE_ACCOUNT_JSON) {
    console.log("[STUB] GSC_SERVICE_ACCOUNT_JSON not configured — skipping GSC ping");
    return null;
  }
  try {
    const { GoogleAuth } = await import("google-auth-library");
    const info = JSON.parse(GSC_SERVICE_ACCOUNT_JSON);
    const auth = new GoogleAuth({
      credentials: info,
      scopes: ["https://www.googleapis.com/auth/webmasters"],
    });
    const client = await auth.getClient();
    const token = await client.getAccessToken();
    const resp = await fetch(
      `https://www.googleapis.com/webmasters/v3/sites/${encodeURIComponent(siteProperty)}/sitemaps/${encodeURIComponent(
        sitemapUrl
      )}`,
      { method: "PUT", headers: { Authorization: `Bearer ${token.token}` } }
    );
    if (!resp.ok) {
      const text = await resp.text().catch(() => "");
      throw new Error(`GSC API ${resp.status}: ${text}`);
    }
    console.log(`GSC sitemap re-submitted: ${sitemapUrl} (property=${siteProperty})`);
    return { ok: true, sitemap: sitemapUrl };
  } catch (e) {
    console.error("GSC ping failed:", e);
    return { error: String(e.message || e) };
  }
}

export async function pingSearchEnginesForSlug(siteUrl, slugs) {
  const base = siteUrl.replace(/\/$/, "");
  const urls = (slugs || []).filter(Boolean).map((s) => `${base}/blog/${s}`);
  // Always include /blog itself so listing pages get re-crawled too.
  urls.push(`${base}/blog`);
  urls.push(`${base}/sitemap.xml`);
  const indexnowResult = await pingIndexnow(urls, siteUrl);
  const gscResult = await pingGscSitemap(`${base}/sitemap.xml`, `${base}/`);
  return { indexnow: indexnowResult, gsc: gscResult, urls };
}
